src/data/list/rotate.lean

```diff
@@ -105,4 +109,241 @@ by rw ← list.take_append_drop (n % list.length (a :: l)) (a :: l) at hl;
+        ←zip_with_distrib_take],
+    all_goals { simp [←h, hn.le] } }
+end
+
+local attribute [simp] rotate_cons_succ
+
+@[simp] lemma zip_with_rotate_one {β : Type*} (f : α → α → β) (x y : α) (l : list α) :
+  zip_with f (x :: y :: l) ((x :: y :: l).rotate 1) =
+  f x y :: zip_with f (y :: l) (l ++ [x]) :=
+by simp
+
+lemma nth_le_rotate_one (l : list α) (k : ℕ) (hk : k < (l.rotate 1).length) :
+  (l.rotate 1).nth_le k hk = l.nth_le ((k + 1) % l.length)
+    (by { cases l, { simpa using hk }, { refine nat.mod_lt _ _, simp }}) :=
+begin
+  cases l with hd tl,
+  { simp },
+  { have : k ≤ tl.length,
+      { refine nat.le_of_lt_succ _,
+        simpa using hk },
+    rcases this.eq_or_lt with rfl|hk',
+    { simp [nth_le_append_right (le_refl _)] },
+    { simpa [nth_le_append _ hk', length_cons, nat.mod_eq_of_lt (nat.succ_lt_succ hk')] } }
+end
+
+lemma nth_le_rotate (l : list α) (n k : ℕ) (hk : k < (l.rotate n).length) :
+  (l.rotate n).nth_le k hk = l.nth_le ((k + n) % l.length)
+    (by { cases l, { simpa using hk }, { refine nat.mod_lt _ _, simp }}) :=
+begin
+  induction n with n hn generalizing l k,
+  { have hk' : k < l.length := by simpa using hk,
+    simp [nat.mod_eq_of_lt hk'] },
+  { simp [nat.succ_eq_add_one, ←rotate_rotate, nth_le_rotate_one, hn l, add_comm, add_left_comm] }
+end
+
+lemma rotate_eq_iff {l l' : list α} {n : ℕ} :
+  l.rotate n = l' ↔ l = l'.rotate (l'.length - n % l'.length) :=
+begin
+  cases l' with x xs,
+  { simp },
+  split,
+  { intro hl,
+    have h : 0 < l.length,
+      { simp [←length_rotate l n, hl] },
+    rw [←hl, rotate_rotate, ←rotate_mod, nat.add_mod],
+    rcases (nat.zero_le (n % l.length)).eq_or_lt with hn|hn,
+    { simp [←hn] },
+    { have : l.length - n % l.length < l.length := nat.sub_lt h hn,
+      simp [nat.mod_eq_of_lt this, nat.add_sub_cancel' (nat.mod_lt _ h).le] } },
+  { intro hl,
+    set l' := (x :: xs) with hx,
+    have : l'.length = (x :: xs).length := congr_arg length hx,
+    rw [hl, this, rotate_rotate, ←rotate_mod, nat.add_mod],
+    rcases (nat.zero_le (n % l'.length)).eq_or_lt with hn|hn,
+    { rw ←hn,
+      simp },
+    { have h : 0 < l'.length := by simp,
+      have : l'.length - n % l'.length < l'.length := nat.sub_lt h hn,
+      rw [nat.mod_eq_of_lt this, nat.sub_add_cancel (nat.mod_lt _ h).le],
+      simp } }
+end
+
+lemma reverse_rotate (l : list α) (n : ℕ) :
+  (l.rotate n).reverse = l.reverse.rotate (l.length - (n % l.length)) :=
+begin
+  rw [←length_reverse l, ←rotate_eq_iff],
+  induction n with n hn generalizing l,
+  { simp },
+  { cases l with hd tl,
+    { simp },
+    { rw [rotate_cons_succ, nat.succ_eq_add_one, ←rotate_rotate, hn],
+      simp } }
+end
+
+section is_rotated
+
+variables (l l' : list α)
+
+def is_rotated : Prop := ∃ n, l.rotate n = l'
+
+infixr ` ~r `:1000 := is_rotated
+
+variables {l l'}
+
+lemma is_rotated.def (h : l ~r l') : ∃ n, l.rotate n = l' := h
+
+lemma is_rotated_iff : l ~r l' ↔ ∃ n, l.rotate n = l' := iff.rfl
```

Comment:
I'm not sure there's any point in either of these - `rw is_rotated` has the same effect as `rw is_rotated_iff`, and `is_rotated.def` amounts to `id`.

Reply:
I use `obtain <n, rfl> := h.def` elsewhere. This type of use prevents destruction of the `h` hypothesis. It's just a shorthand that of course could be emulated via some iff rewrite, cloning of the hypothesis, use of defeq, etc.

Reply:
If your problem is simply that `obtain` is destroying your hypothesis, you can use `obtain ⟨x, hx⟩ := id h,`

Reply:
OK. I will remove these tautologies. But if someone changed the defn of the Prop to be defined via an inductive, then it'd be nice to have them. The inductive definition leads to better computation, I think. But I haven't had to use computability yet, so I haven't switched the defn.